src/backend/optimizer/plan/createplan.c

```diff
@@ -784,9 +784,11 @@ create_join_plan(PlannerInfo *root, JoinPath *best_path)
 
 	/* CDB: if the join's locus is bottleneck which means the
 	 * join gang only contains one process, so there is no
-	 * risk for motion deadlock.
+	 * risk for motion deadlock. But prefetch_inner may still be necessary in the
+	 * case of HashJoin with SharedScan, the writing part of which is always in
+	 * the inner. So do not override it in the case of HashJoin.
 	 */
-	if (CdbPathLocus_IsBottleneck(best_path->path.locus))
+	if (CdbPathLocus_IsBottleneck(best_path->path.locus) && !IsA(plan, HashJoin))
```

Comment:
Maybe move this condition in the exact join type function? Overwriting of `prefetch_joinqual` and `prefetch_qual` seems to be useless and may be omitted after e0cca545510322805731d5eb723f512df6b25f29. Moreover, the previous condition looks contradicting with this block of code. We enable prefetch by one condition and disable by another.

Reply:
if delete prefetch_joinqual and prefetch_qual, then moving the condition to the place where the nested loop and merge join were created makes sense.

disabling prefetch is necessary for nested loop without a patch to fix the reader on slice 0
is also necessary for merge join, in which, without disabling prefetch, inconsistency with travel in shareinput_walker and in the executor may occur.